src/Microsoft.AspNet.Hosting/Internal/WebApplication.cs

```diff
@@ -98,14 +127,14 @@ public virtual IApplication Start()
             _applicationLifetime.NotifyStarted();
             logger.Started();
 
-            return new Application(ApplicationServices, Server.Features, new Disposable(() =>
-            {
-                logger.Shutdown();
-                _applicationLifetime.StopApplication();
-                Server.Dispose();
-                _applicationLifetime.NotifyStopped();
-                (_applicationServices as IDisposable)?.Dispose();
-            }));
+            return new Disposable(() =>
```

Comment:
`WebApplication` and `WebApplicationInstance` doesn't resolve the usage case illustrated by `StartupConfigureAddresses` unless we add some sort of `ConfigureWebApplicationInstance` callback. We need to finish this up so I'll open a new issue tracking the discussion here and a few other items once this PR is done.